### ISSUE TITLE
Memphis - Issue with billing locked

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1671]
In GitLab by @kdjstudios on Apr 3, 2020, 09:13

**Submitted by:**  "Laura Duckworth" <laura.duckworth@answernet.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2020-04-02-62159

**Server:** Internal

**Client/Site:** Memphis

**Account:** NA

**Issue:** 

I am trying to process y billing and it will not allow me to upload my file. I keep getting a message telling me its locked by user. Can you assist me with having this cleared so that I can process my billing for the month.

![image](/uploads/7b6b0b2675fee62ef1618617941c3825/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 3, 2020, 09:17

I wrote:

> May you please log out and log back in and attempt to process your billing cycle.
> 
> That locked message means another user is also trying to perform steps on the same billing cycle; or that one user has multiple windows/tabs open trying to perform steps on the same billing cycle. The only way to unlock a billing cycle is to ensure that only one user is processing the billing cycle at a time.

@amishra @ranand @Robleski @tim.traylor 

We will need to know why it is not populating with the username and fix that. Please let us know if this bug was due to a release, which release ticket it was.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 3, 2020, 09:50

@kdjstudios @tim.traylor 

Thank you for the update.

* You can click on the billing cycle button again then they will work normally.
* Let us know if still facing the same issue.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 3, 2020, 10:44

@amishra @ranand @tim.traylor @Robleski 

First, @amishra I am completely confused with your response. How does that address the issue that the username is not populating. I will presume we need to know that in order to log out that person, to unlock the billing cycle correct? 

Laura wrote:

>  I've logged out multiple times reset my computer and I still keep getting the message.

I wrote:

> Which Site and Billing cycle are you attempting to process? Also have you tried reverting the billing cycle and starting over?

Laura wrote:

> for the Memphis sit and I'm trying to create this 4/1/2020 bill cycle and upload the file  for billing and it gets stuck there. I have reverted and cxl the billing.

I wrote:

> I was just able to go into the Memphis site and process the "No Usage" step on the Master billing cycle without issue or error. I have also Reverted this too, so you should be back ready to upload the usage. It does not appear there are any other users locking the billing cycle. I would recomend you check the billing file you are uploading an possibly rerun the report in case that is where the error is coming from. 
> 
> May you please reattempt to process the billing cycle. If you are still having issues, lets setup a zoom meeting so we can further look into this. 

Laurel wrote:

I wrote:

> May you please provide the billing file you are attempting to upload?

Laura wrote:

> Here you go

[Memphis_April.csv](/uploads/37294abf2753221b2ddc3c17be329f59/Memphis_April.csv)


What is going on?!?!?! I waited over 20 minutes to upload the usage file!!! And it was still processing. I finally gave up.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 3, 2020, 15:37

@amishra @ranand 

Please look into this and have it fixed and working by Monday morning. Updating priority since this is now preventing operations from completing their billing. Please also explain the above questions on why there is no user populating and which release/update caused this to occur.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 5, 2020, 23:51

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 6, 2020, 24:47

@kdjstudios @tim.traylor 

Thank you for the update.

We looked into this and our findings are described below:

### Upload Usage File

* As you mentioned above the usage file has not valid as you can see there were no valid usages data.
* Please upload the correct usage file and it will work correctly.
* Also we have tested with upload the usage file on the dev server for Memphis, TN site.
* **Upload the file** : **`~ 35 sec`**
* **Apply the billing cycle** : **`3-4 min`**
* **Finalize the billing cycle** : **`50 sec`**

>. Please also explain the above questions on why there is no user populating and which release/update caused this to occur.

* We need to time to figure out why the username is not populating correctly.I will update you as soon as possible.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 6, 2020, 08:31

I wrote:

> It would appear that is the wrong billing file you are trying to upload. I believe I saw in a separate email that you were able to process your billing cycle once you used the correct billing file. May you please confirm that we can close this ticket?

laura wrote:

> yes you can close it, thanks

@amishra @Robleski 

Thanks for the updates. I do have one more question now. I could have sworn we were able to show a corresponding error if the wrong file was trying to be uploaded. Such there were missing headers, etc. Did that get removed or was that just never completed. I am trying to find the ticket now and will post and update once I am able to find it.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 14, 2020, 24:54

@kdjstudios @tim.traylor 

Thank you for the update.

We looked into this and our findings are described below:

* The main issue was of the ticket, unable to upload to the usage file
  * While creating the transaction we find the account from a resource id
  * they had missed the resource id column on the cdr records and the resource id column is not present on the usage file.

* As per the current functionality, we are handling the code, if ruby does not convert the usage file. 
* All the header dynamic based on the site Switch. All usage file headers are different.

* Let me know if we need to improve the usages upload process. what case we need to handle.

**`NOTE: THE MAIN ISSUE WAS RESOLVED THIS TICKET THEN WE CAN REPRIORITIZE THIS TICKET.`**

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 15, 2020, 08:55

Marking this for discussion to determine if we should be displaying some form of error or warning when a file is uploaded that does not match the expected file "format".

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on May 17, 2020, 23:58

@kdjstudios  @Robleski 

Thanks for the update.
* We dig into the issue and tried to regenerate the notice without username, but couldn't find a scenario for which the locking mechanism is not working.

* We have validated the file format for CSV which accepts CSV format only. 

* In regards to validating the headers of CSV, it difficult to achieve that. As we are not storing the switch headers of the CSV anywhere and every switch has their own set of CSV headers. Please suggest your thought on how we can validate the CSV file header.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 18, 2020, 11:03

Thanks @rranjan for the update.

@Robleski I think we will want to review this further. 

In regards to validating the headers of CSV, it difficult to achieve that. As we are not storing the switch headers of the CSV anywhere and every switch has their own set of CSV headers. Please suggest your thought on how we can validate the CSV file header.

> This seems odd, as how are we currently matching the imported values during the upload if we are not "matching" them via the column headers? I know once we upload them we should be saving them into the 55000 and 7000 codes. I do not see any other way of doing this. If we actually are matching them via the headers I would think during the upload process a simple counter to see if we end up matching all headers or if some headers end up being missed then we can display an warning/error of which headers were not part of the file. Would that makes sense or is there a better way?

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on May 21, 2020, 04:56

@kdjstudios @Robleski 

We can implement the above given suggestion. We have static/hard-coded headers for each switch, we can match them at the time uploading.

Please find the estimate to validate the CSV header for all current active switches.


## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|R-Small|4.5|Concepts and Implementation Well Understood|5.0|High Impact|2.0|7.0|

##### Estimate Notes:
* Risk areas identified: `Upload usage process of the Billing Cycle`
* Assumptions made: `It will be done for all the active switches only.`
* Work plan: 
  * `Make a list of headers for all the active switch`
  * `Match static headers of the switch with the file headers. If any required header is missing, it will display the notice to the user`
* Test cases:     
  * `For the missing header, it will display the static error message`
  * `For the valid header CSV, it will upload the usage as usual`


**Please provide the verbiage of the notice in case of missing/invalid headers.**

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 21, 2020, 13:16

I believe this should be the notice. @Robleski please review and confirm. 

UI Notice:

> The uploaded file was missing the following headers:
> * Header 1
> * Header 2
> Please ensure you uploaded the correct usage file. You may want to revert and re-upload or you may continue to process your billing without the usage for the above.

Email Notice:

* To: site email, user email
* BCC: apperrors@sahosted.com
* Body:
> The uploaded file was missing the following headers:
> * Header 1
> * Header 2
> Please ensure you uploaded the correct usage file. You may want to revert and re-upload or you may continue to process your billing without the usage for the above.

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on May 22, 2020, 07:03

approved for Avaaz to start work

**Comment by kdjstudios** [comment]
In GitLab by @amishra on May 27, 2020, 23:17

assigned to @rranjan and unassigned @amishra

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on May 28, 2020, 09:24

In Avaaz testing.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on May 29, 2020, 08:11

@kdjstudios  @Robleski 

Thanks for the update.

We have completed the development and setup branch on the dev server for your testing. Please test and give us your feedback.

*  For from email we are using `support@sabilling.com` do we need to change?

*  Also provide the email subject.

**NOTE:-** Please test with the **demo site**.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 29, 2020, 09:23

@rranjan @Robleski 

The UI error looks to be working as expected. Lets use the following for the subject line on the email error: 

"User [Username] uploaded usage file for [BillingCycleName] for Site [Sitename] with missing metric headers/columns"
> Example: "User kyjohnson uploaded usage file for 2/2/20 Master for Site Billings, MT with missing metric headers/columns"

However, I was not able to continue processing after receiving the error in the UI. This kept me on the "upload" step.  I should have been advanced to the "missing accounts" or "apply" steps accordingly. NOTE: this is because the user can always "revert" a billing cycle, so there is no need to stop them from processing. If needed to display this error in a popup that is fine too.

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on Jun 1, 2020, 07:16

In AN Acceptance Testing. Greg to follow up with Kyle

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 1, 2020, 10:17

@Robleski @ranand @rranjan 

This is in development for Aavaz, not waiting on Answernet. Please see http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1774#note_59453. Unless someone did not update the labels to specify which server this is on and needed to be re-tested.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 2, 2020, 08:30

@kdjstudios  @Robleski 

Thanks for the update.

> "User \[Username\] uploaded usage file for \[BillingCycleName\] for Site \[Sitename\] with missing metric headers/columns"

> Example: "User johnson uploaded usage file for 2/2/20 Master for Site Billings, MT with missing metric headers/columns"

![image](/uploads/8ea469304b45d3d671a88c18a83d460d/image.png)

We have changed accordingly.


> However, I was not able to continue processing after receiving the error in the UI. This kept me on the "upload" step. I should have been advanced to the "missing accounts" or "apply" steps accordingly. NOTE: this is because the user can always "revert" a billing cycle, so there is no need to stop them from processing. If needed to display this error in a popup that is fine too.


We can't proceed with the upload process with the invalid header usage file because of the following reason:-

*  There will be some missing header which could be used in the activity calculation. 

*  There will be some essential header which could be missing for processing the usage file. **i.e - resource_id,client_id, Account number**


Do you want to test this again?

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 2, 2020, 10:18

@rranjan @ranand @amishra @Robleski 

We can't proceed with the upload process with the invalid header usage file because of the following reason:-
> How can we not proceed? We already/currently have been allowing ops to proceed if they have been missing? We need to allow for exactly what we are already doing. I am so confused with what else was changed that was not requested to be changed.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Jun 3, 2020, 01:36

@kdjstudios @Robleski 

* We allow ops to proceed only in the case of invalid/incorrect resource_id value(the primary column value). If we don't find any account with the given resource_id then we add it to the missing accounts list.
* Here the case is different and there could be several cases to handle-:
  * Primary Column is missing: ie. the column through which we find the account is missing (eg. resource_id, account_number)
  * Secondary Column is missing: the column which is being used in calculating the usage activities. Eg. The one we faced in #1824, because of Work:Count, Work:Time, Work:Unit missing headers there were missing revenue in several activities and the user had to revert the billing cycle.
* When we generally raise an error, it halts the process then & there and it does not go to the next step of the process.

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on Jun 3, 2020, 07:01

Will review at Thursday zoom.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 4, 2020, 08:52

This was discussed today. We will be allowing for the user to continue to process their billing cycle after showing the error/warning about the missing file headers.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 9, 2020, 02:31

@kdjstudios  @Robleski 

We have set up a branch on the dev server for your testing.


Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 9, 2020, 08:16

@rranjan @Robleski 

Who is testing this? I tried to upload a usage file and received the following error. There should never be a reason I receive an error like this during testing if your team had correctly tested this.

![image](/uploads/ea0bf90a17fd287b90b1ae7d71350ab4/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 10, 2020, 08:53

@amishra @rranjan @ranand @Robleski

Please let us know when this is ready for our testing.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 11, 2020, 06:08

@kdjstudios @Robleski 

Although we have fixed the resource_id issue.

We have the following query:- 

According to the current functionality of SAB if the resource_id column is `nil(blank)` it finds the account with nil resource id. We have a lot of accounts in SAB whose resource_id is nil. So, it find first account with nil resource_id and uploads the usage on it.

So, What we need to do in this scenario.?

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on Jun 11, 2020, 08:33

Back in Avaaz dev.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 11, 2020, 08:39

If the column is missing then display it on the error/warning that the column is missing.

If the Resource ID is missing in the file/row, please display the row(s) on the exceptions/missing accounts page.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 12, 2020, 08:28

@kdjstudios  @Robleski 

Thanks for the update.

If the column is missing then display it on the error/warning that the column is missing.

If the Resource ID is missing in the file/row, please display the row(s) on the exceptions/missing accounts page

> We have developed it accordingly.

 We have added these changes for the all **Internal** and **External** Active Switch version. It took 4 additional hours to develop this.

*  V2.0.1
*  V2.0.2
*  V2.0.3
*  V1.2.1
*  V1.2.2
*  V1.2.3 etc.


We set up a branch on the merge2 server for your testing.


Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 15, 2020, 09:17

@rranjan @Robleski @amishra @ranand 

I just want to confirm you developed this for all active switches internal and external, such as the latest VCC 1 and 2 versions, Speed Ez, Towne, etc.? I ask this because your statement above appears to state you developed this for inactive versions of just the VCC switches.

I notice the email is not formatted, as the "Please ensure...." should be on a new line. Please fix. 

![image](/uploads/c3088a8235e0bd6c91f611d230f17a83/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 16, 2020, 02:26

@kdjstudios  @Robleski 

Thanks for the update.

> I just want to confirm you developed this for all active switches internal and external, such as the latest VCC 1 and 2 versions, Speed Ez, Towne, etc.? I ask this because your statement above appears to state you developed this for inactive versions of just the VCC switches.

 Fixed Header Missing Functionality

  *  Active VCC V1 Version
  *  Active VCC V2 Version
  *  Speed Ez(Startel)
  *  Towne(Pinnacle)
  *  Laser Answering Service(Infinity)

 Fixed Resource ID issue

*  Active VCC V1 Version
*  Active VCC V2 Version

**Note:-** For NCSM, Speed Ez, and  Towne,Laser Answering Service resource_id issue was not occurring. So we don't need to fix it.

> I notice the email is not formatted, as the "Please ensure...." should be on a new line. Please fix.

 We looked into it and find that it is showing on a new line.
![image](/uploads/d5ec31b0c0dc2850b70ec14152b81168/image.png)

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 16, 2020, 08:45

@rranjan @Robleski @amishra @ranand 

The email is showing as not on a new line in Outlook and the web-client, so I presume your email client is doing some additional formatting. 

I tried to retest this today and received the following error when trying to get to the billing cycle page. Who tested this!? How did you determine this was ready for us, when the page does not even load?

![image](/uploads/ecbf6d78f6e5a5fe0c6214f13563f050/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 17, 2020, 07:11

@kdjstudios  @Robleski 

Thanks for your feedback

>The email is showing as not on a new line in Outlook and the web-client, so I presume your email client is doing some additional formatting.

 we looked into it and find that it's showing on the new line for Outlook but we don't test for web-client because we don't have an account there. Although we have added line break.

![image](/uploads/47699e8083076303bcaf31702ee3c77b/image.png)

> I tried to retest this today and received the following error when trying to get to the billing cycle page. Who tested this!? How did you determine this was ready for us when the page does not even load?

Yesterday, I was working late on the merge2 server so you got the error.

we have set up a branch on the merge2 server for your testing. Please test it and give us your feedback.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 17, 2020, 08:59

@rranjan @ranand @amishra @Robleski 

This is still not working. I uploaded a blank CSV file [BlankCSV.csv](/uploads/b3917a5e7ccb6b7fe0b0f7422959e3ed/BlankCSV.csv) to the Demo site on Merge2. So it should say all headers are missing. I received no notification that any headers were missing. Please **DO NOT** push this to our testing until you have fully tested this.  

![image](/uploads/50b82ca13da6fa395f81b9859f19edd1/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 19, 2020, 24:18

@kdjstudios @Robleski 

Thanks for your feedback.

We have tested all the test cases mentioned below. Please let us know if we missed any test cases.
  
[Test Cases](http://gitlab.aavaz.biz/AnswerNet/SABilling/wikis/Test-Cases-For-Header-Missing-%231774-:-VCC-V2-&-NCSM)

Also, We have set up a branch on the merge2 server for your testing.

NOTE:- While development we encounter an existing issue and fixed it. It took 2 additional hours, the issue is mentioned in test cases no- 7 and 12.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 19, 2020, 09:41

Marking this for annoucment.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 27, 2020, 04:53

@kdjstudios  @Robleski 

We have successfully released on both of the production servers.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 30, 2020, 10:19

closed
